### PR TITLE
Allow MonoGCBridgeSCC objects to have an empty User Peer list

### PR DESCRIPTION
**Pull request by xmcclure**
Mono passes gc_cross_references a description of a graph. Each node in the graph has a list of User Peer objects which should be collected together in the upcoming GC. This list is required by both the spec and the current code to be nonempty. To enable new optimizations, the runtime team wants to modify the spec to allow the list to be empty. This patch makes that possible by creating a new temporary object of type GCUserPeer when a MonoGCBridgeSCC with an empty list is seen.

I think this patch should not be merged until tests with Mono have shown that this can actually produce a performance improvement. In the meantime you can test this with https://github.com/xmcclure/mono/tree/andi-phantom-test-1 and https://github.com/xmcclure/monodroid/commits/andi-phantom-test-1 .

### CODE REVIEW

**Comment by xamarin-release-manager** [comment]
Hello! I'm the build bot for the Mono project.

I need approval from a Mono team member to build this pull request. A team member should reply with "approve" to approve a build of this pull request, "whitelist" to whitelist this and all future pull requests from this contributor, or "build" to explicitly request a build, even if one has already been done.

Contributors can ignore this message.

File: src/monodroid/jni/monodroid-glue.c [hunk 1]
```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
...
-#if DEBUG
-	MonoClass *klass;
-	klass = mono.mono_object_get_class (obj);
-#endif
 
-	mono.mono_field_get_value (obj, bridge_info->handle, &handle);
 	java_class = (*env)->GetObjectClass (env, handle);
 	add_method_id = (*env)->GetMethodID (env, java_class, "monodroidAddReference", "(Ljava/lang/Object;)V");
 	if (add_method_id) {
-		mono.mono_field_get_value (reffed_obj, bridge_info->handle, &reffed_handle);
```
Comment (xmcclure) [inline src/monodroid/jni/monodroid-glue.c]: Notice there is a pretty big bug here which this patch fixes. The same bridge_info is used for obj and reffed_obj. If obj is an Object and reffed_obj is a Throwable, this could lead to Something Very Bad happening.

File: src/monodroid/jni/monodroid-glue.c [hunk 2]
```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
...
-			if (bridge_info != NULL && add_reference (env, scc->objs [scc->num_objs-1], bridge_info, scc->objs [0])) {
-				mono.mono_field_set_value (scc->objs [scc->num_objs-1], bridge_info->refs_added, &ref_val);
+			MonoGCBridgeSCC *first = scc->objs [0];
+			MonoGCBridgeSCC *prev = first;
+
+			/* start at the second item, ref j from j-1 */
+			for (int j = 1; j < scc->num_objs; j++) {
+				MonoGCBridgeSCC *current = scc->objs [j];
+
+				add_reference_mono_object (env, prev, current);
```
Comment (xmcclure) [inline src/monodroid/jni/monodroid-glue.c]: Notice a pretty bad problem with both the original and patched versions of this line: add_reference can fail, and if that happens the reference protections for the SCC will be incomplete. This could lead to objects being collected which are still in use.

File: src/monodroid/jni/monodroid-glue.c [hunk 3]
```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
...
+		mono.mono_field_get_value (target.obj, (*bridge_info)->handle, handle);
+	} else {
+		*handle = target.jobj;
+	}
+	return TRUE;
+}
+
+#if DEBUG
+// Allocate and return a string describing a target
+static char *describe_target (AddReferenceTarget target) {
```
Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: `describe_target` should go on a new line.

File: src/monodroid/jni/monodroid-glue.c [hunk 4]
```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
...
+		free (description);
+		free (reffed_description);
+	}
+#endif
+
+	return success;
+}
+
+// Create a target
+AddReferenceTarget mono_object_target (MonoObject *obj) {
```
Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: This should be formatted as:

    static AddReferenceTarget
    mono_object_target (MonoObject *obj)
    {
        ...
    }

Comment (xmcclure) [inline src/monodroid/jni/monodroid-glue.c]: the mono_ is because this is a MonoObject. there is no such thing as a monodroid_object. maybe i could prefix both _target methods with make_ so it is clear the mono_ is not a namespace ?

Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: My larger concern is that monodroid doesn't control mono, and nothing prevents mono from creating a `mono_object_target()` method in the future, at which point we'd have potential conflicts/confusion.

As such, a `from_` prefix might be fine (`from_mono_object()`), or a `get_*_target()` pattern (`get_mono_object_target()`, `get_jobject_target()`).

File: src/monodroid/jni/monodroid-glue.c [hunk 5]
```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
...
+	AddReferenceTarget result;
+	result.is_mono_object = TRUE;
+	result.obj = obj;
+	return result;
+}
+
+// Create a target
+AddReferenceTarget jobject_target (jobject jobj) {
```
Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: Formatting: this should be `static`, return type should go on a separate line, `{` on a new line.

File: src/monodroid/jni/monodroid-glue.c [hunk 6]
```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
...
+	AddReferenceTarget result;
+	result.is_mono_object = FALSE;
+	result.jobj = jobj;
+	return result;
+}
+
+// Extract the root target for an SCC. If the SCC has bridged objects, this is the first object. If not, it's stored in temporary_peers.
+// Must pass in the JNI information necessary to interact with temporary_peers.
+static AddReferenceTarget scc_target (MonoGCBridgeSCC **sccs, int idx, JNIEnv *env, jobject temporary_peers, jmethodID arraylist_get, int *temporary_peer_indices)
```
Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: `scc_target (...)` should go on a new line.

File: src/monodroid/jni/monodroid-glue.c [hunk 7]
```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
...
+	MonoGCBridgeSCC *scc = sccs [idx];
+	if (scc->num_objs > 0)
+		return mono_object_target (scc->objs [0]);
+
+	jobject jobj = (*env)->CallObjectMethod (env, temporary_peers, arraylist_get, temporary_peer_indices [idx]);
+	return jobject_target (jobj);
+}
+
+// Must call this on any AddReferenceTarget returned by scc_target once done with it
+static void release_scc_target (JNIEnv *env, AddReferenceTarget target) {
```
Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: `release_scc_target (...)` should go on a newline, and `{` should go on a new line.

File: src/monodroid/jni/monodroid-glue.c [hunk 8]
```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
...
+add_reference_mono_object (JNIEnv *env, MonoObject *obj, MonoObject *reffed_obj)
+{
+	return add_reference (env, mono_object_target (obj), mono_object_target (reffed_obj));
+}
+
+// Clear a local reference which might be NULL // FIXME: Is it safe to just call DeleteLocalRef with NULL?
+static void
+delete_java_lref (JNIEnv *env, jobject jobj)
+{
+	if (jobj)
```
Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: This check shouldn't be necessary.

Additionally, I'm not sure this function is "useful", and most current uses should instead use `(*env)->DeleteLocalRef(env, ...)` instead.

File: src/monodroid/jni/monodroid-glue.c [hunk 9]
```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
...
+}
+
 static void
 gc_prepare_for_java_collection (JNIEnv *env, int num_sccs, MonoGCBridgeSCC **sccs, int num_xrefs, MonoGCBridgeXRef *xrefs)
 {
-	MonoGCBridgeSCC *scc;
-	int i, j, ref_val;
+	/* Some SCCs might have no IGCUserPeers associated with them, so we must create one */
+	jobject arraylist = NULL, gcuserpeer = NULL;
+	jmethodID gcuserpeer_constructor = NULL, arraylist_get = NULL, arraylist_add = NULL;
```
Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: `gcuserpeer_constructor`, `arraylist_get`, and `arraylist_add` should be "promoted" into global variables, looked up in `JNI_OnLoad()`. We don't want to have to perform these class and jmethodID lookups potentially during every collection, as that's extra avoidable overhead:

    /* around line 730... */
    static jobject GCUserPeer_class; /* gref */
    static jmethodID GCUserPeer_ctor;

    static jobject ArrayList_class; /* gref */
    static jmethodID ArrayList_ctor;
    static jmethodID ArrayList_add;
    static jmethodID ArrayList_get;

Comment (xmcclure) [inline src/monodroid/jni/monodroid-glue.c]: @jonpryor: Promoting them into global variables means reserving a gref for each of `arraylist` and `gcuserpeer` and never letting go of them. That's acceptable, then?

Are class/methodid lookups expensive? I notice we're calling GetMethodID on *every* object we add a reference for, currently…

Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: > That's acceptable, then

It's only two more, so yes.

> Are class/methodid lookups expensive

Expensive enough that we want to avoid them when possible. See also: https://bugzilla.xamarin.com/show_bug.cgi?id=32126

>  I notice we're calling GetMethodID on every object we add a reference for, currently

Indeed! This is also something [addressed in Java.Interop](https://github.com/xamarin/java.interop/blob/9fbade1/src/java-interop/java-interop-gc-bridge-mono.c#L845-L859), and we can worry about it in the eventual Java.Interop/xamarin-android GC bridge reunion.

File: src/Mono.Android/java/mono/android/GCUserPeer.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package mono.android;
+
+class GCUserPeer {
```
Comment (jonpryor) [inline src/Mono.Android/java/mono/android/GCUserPeer.java]: shouldn't this implement `IGCUserPeer`?

Comment (xmcclure) [inline src/Mono.Android/java/mono/android/GCUserPeer.java]: It could, it doesn't need to. We load monodroidAddReference off of the class directly using GetMethodID. In fact, monodroid-glue doesn't use IGCUserPeer at all. We fetch a gref to it.

Would you *prefer* I have `GCUserPeer` implement `IGCUserPeer`?

File: src/monodroid/jni/monodroid-glue.c [hunk 11]
```diff
@@ -1278,91 +1278,249 @@ gc_is_bridge_object (MonoObject *object)
...
+	mono_bool is_mono_object;
+	union {
+		MonoObject *obj;
+		jobject jobj;
+	};
+} AddReferenceTarget;
+
+// These will be loaded as needed and persist between GCs
+// FIXME: This code assumes it is totally safe to hold onto these GREFs forever. Can mono.android.jar ever be unloaded?
+static jobject   jni_arraylist, jni_gcuserpeer;
```
Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: Convention is:

1. Prefix based on the Java class name (excluding package), preserving case. **Good**: `ArrayList`. **Bad**: `arraylist`.

2. Underscore.

3. "Member name": the field/method name, or `ctor` for constructors, or `class` for `jclass` values.

These are also usually done one per line, organized by type:

Thus:

    static jclass     ArrayList_class;
    static jmethodID  ArrayList_ctor;
    static jmethodID  ArrayList_get;
    static jmethodID  ArrayList_add;

    static jclass     GCUserPeer_class;
    static jmethodID  GCUserPeer_ctor;

File: src/monodroid/jni/monodroid-glue.c [hunk 12]
```diff
@@ -1278,91 +1278,249 @@ gc_is_bridge_object (MonoObject *object)
...
+	result.is_mono_object = FALSE;
+	result.jobj = jobj;
+	return result;
+}
+
+// See note in gc_prepare_for_java_collection
+static int
+scc_get_stashed_index (MonoGCBridgeSCC *scc)
+{
+	assert (scc->num_objs < 0);
```
Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: We want `num_objs` to be *negative*? That doesn't seem right...

Additionally, we should strive to have better asserts. (Yes, I'm not doing good here, glass houses and all that...)

Specifically, we should have a useful error message in here:

    assert ((scc->num_objs < 0) || !"Description for why we're asserting");

File: src/monodroid/jni/monodroid-glue.c [hunk 13]
```diff
@@ -1278,91 +1278,249 @@ gc_is_bridge_object (MonoObject *object)
...
-		MonoJavaGCBridgeInfo    *bridge_info = NULL;
-		/* start at the second item, ref j from j-1 */
-		for (j = 1; j < scc->num_objs; j++) {
-			bridge_info = get_gc_bridge_info_for_object (scc->objs [j-1]);
-			if (bridge_info != NULL && add_reference (env, scc->objs [j-1], bridge_info, scc->objs [j])) {
-				mono.mono_field_set_value (scc->objs [j-1], bridge_info->refs_added, &ref_val);
-			}
-		}
-		/* ref the first from the last */
+	/* Basic setup before looking at xrefs */
```
Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: Not a very descriptive comment. The prior comment of `add java refs for items in the list which reference each other` was more descriptive. Either remove this comment (preferred?) or provide a better one.

Related: you might want to use more "helper" functions for each case:

    for (int i = 0; i < num_sccs; i++) {
        MonoGCBridgeSCC *scc = sccs [i];
        if (scc->num_objs > 1) {
            setup_whatever_refs (...);
        } else if (scc->num_objs <= 0) {
            setup_other_refs (...);
        }
    }

Another advantage to doing this is that it forces us to create names. What does it *mean* when `scc->num_objs <= 0`? That should have some kind of name, which we can then use consistently (or in only one place, but at least we have a named construct!).

File: src/monodroid/jni/monodroid-glue.c (hunk 12 above)
Comment (xmcclure) [inline src/monodroid/jni/monodroid-glue.c]: Although this is outside the scope of this patch, if this is a thing you're looking for I'd suggest making an assert () variant which actually takes a description string.

File: src/monodroid/jni/monodroid-glue.c [hunk 14]
```diff
@@ -1278,92 +1278,271 @@ gc_is_bridge_object (MonoObject *object)
...
+	return target_from_jobject (jobj);
+}
+
+// Must call this on any AddReferenceTarget returned by target_from_scc once done with it
+static void
+target_release (JNIEnv *env, AddReferenceTarget target)
+{
+	if (!target.is_mono_object)
+		(*env)->DeleteLocalRef (env, target.jobj);
+}
```
Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.c]: Should we set `target.jobj=NULL`?

Comment (xmcclure) [inline src/monodroid/jni/monodroid-glue.c]: Hm, I do not think so, any more than DeleteLocalRef should set it to NULL. This function is only used when the target object is done with (and currently is set in only one well-defined place).

Comment (xmcclure) [inline src/monodroid/jni/monodroid-glue.c]: Wait, hold on: No, we should not set it to NULL, because target is not being passed by reference. "target" is here a function temporary.
